### ISSUE TITLE
[operator] look into ansible "turbo" mode and removing ansible loops

### ISSUE THREAD

**Issue by jmazzitelli** [kiali/kiali#6032]
Yes, turbo mode :) Blast from the 80s past, there may not be a "turbo" button on your PC anymore, but it looks like Ansible has an experimental feature "turbo mode". We should look into this and see how it affects the Kiali operator.

see: https://github.com/ansible-collections/kubernetes.core/blob/main/docs/ansible_turbo_mode.rst

Word is the main problems appear to be related to memory consumption (wrt turbo mode) and is still considered experimental.

But I was also told (and can confirm based on https://github.com/kiali/kiali-operator/pull/636) that refactoring/removing ansible loops provides potentially even more performance gains. So that should also be considered as part of work on this issue.

PRs for this issue that help speed up the operator:
* https://github.com/kiali/kiali-operator/pull/636
* https://github.com/kiali/kiali-operator/pull/637
* https://github.com/kiali/kiali-operator/pull/639

backport PRs:
* https://github.com/kiali/kiali-operator/pull/638

**Comment by jmazzitelli** [comment]
While doing this, if the experimental nature of this turbo mode is not something we want, we should look at using the async/poll/wait features of the `k8s` task - that supports the ability to process things asynchronously. I'm not sure if it would help much, but we should look into it.

**Comment by jmazzitelli** [comment]
> While doing this, if the experimental nature of this turbo mode is not something we want, we should look at using the async/poll/wait features of the `k8s` task - that supports the ability to process things asynchronously. I'm not sure if it would help much, but we should look into it.

Well, we can forget that - `async` is not supported with the `k8s` task:

Run this playbook:

```yaml
- hosts: localhost
  gather_facts: no
  vars:
    ansible_python_interpreter: python3
  collections:
  - kubernetes.core

  tasks:
  - name: Does async work?
    async: 60
    poll: 0
    k8s:
      state: absent
      kind: ConfigMap
      namespace: default
      name: foo
```

and you get:

```
TASK [Does async work?] ***************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "async is not supported for this task."}
```

**Comment by jmazzitelli** [comment]
I did a VERY quick test. I installed the necessary cloud.common ansible collection and set the environment variable in the kiali operator playbooks:

```diff
$ git diff
diff --git a/playbooks/kiali-deploy.yml b/playbooks/kiali-deploy.yml
index 75b48d0..e525e9b 100644
--- a/playbooks/kiali-deploy.yml
+++ b/playbooks/kiali-deploy.yml
@@ -2,6 +2,8 @@
   gather_facts: no
   collections:
   - kubernetes.core
+  environment:
+    ENABLE_TURBO_MODE: 1
   tasks:
 
   - debug:
diff --git a/playbooks/kiali-remove.yml b/playbooks/kiali-remove.yml
index 71dcbb5..76b00d4 100644
--- a/playbooks/kiali-remove.yml
+++ b/playbooks/kiali-remove.yml
@@ -2,6 +2,8 @@
   gather_facts: no
   collections:
   - kubernetes.core
+  environment:
+    ENABLE_TURBO_MODE: 1
   tasks:
 
   - ignore_errors: yes
diff --git a/requirements.yml b/requirements.yml
index 8c71447..cb75d58 100644
--- a/requirements.yml
+++ b/requirements.yml
@@ -38,4 +38,5 @@ collections:
   version: 2.3.1
 - name: operator_sdk.util
   version: 0.4.0
+- name: cloud.common
```

I then ran `make -e OPERATOR_PROFILER_ENABLED=true -e CLUSTER_TYPE=minikube run-operator-playbook` to get the create/delete playbooks run on my local machine where I already have ansible installed. I ran it three times with turbo mode on, and three times with it disabled.

I looked at the time it took for the "remove" playbook to run (this is when the Kiali CR is deleted and Kiali is uninstalled and all its resources are removed).

In short, turbo mode shaved 4.63 seconds from the run (a 9.5% speed up).  Without turbo mode, the remove playbook took on average 48.67s, with turbo mode it took on average 44.04s.

The `default/kiali-remove : Delete Kiali resources` task (found in roles/default/kiali-remove/tasks/main.yml) was the main culprit which took up more than half the time. On average, without Turbo mode that task took 25.17s and with Turbo mode it took 22.39 seconds (shaved 2.78s, an 11% speed up)

So it looks like turbo mode will give roughly a 10% speed up.

**Comment by jmazzitelli** [comment]
I ran another quick test, this time with both the deploy and remove playbooks with turbo mode disabled (i.e. the current situation). Running both with the same test procedure as before results in an average time of 52.97s. When both playbooks had turbo mode enabled (the previous testing shows this), the average time was 44.04s. The time savings was 8.93s for a speed up of 16.86%.

So, in short, this quick set of tests shows we can expect something around 17% speed up with turbo mode. That's not nothing, but it isn't earth-shattering either.

**Comment by jshaughn** [comment]
If there is no downside why not take the small win...

**Comment by jmazzitelli** [comment]
There is also this: https://www.ansible.com/blog/whats-new-in-the-ansible-content-collection-for-kubernetes-2.0, specifically "Scenario: Apply multiple templates in one task" - this speaks directly to the loop the Kiali operator has when it creates/deletes resources (where the bulk of the time is spent).

I'll look further into this and turbo mode - I do not know what kind of downsides turbo mode would introduce.

**Comment by jshaughn** [comment]
> I do not know what kind of downsides turbo mode would introduce.

Yeah, what makes me a little nervous is that it is a "mode".  That usually means you are giving something up, because otherwise why not just make these standard perf improvements.

**Comment by jmazzitelli** [comment]
I ran molecule tests on minikube with all the latest PRs and the results were:

```
              accessible-namespaces-test... success [8m 27s]
     affinity-tolerations-resources-test... success [3m 39s]
                      config-values-test... success [2m 44s]
                  default-namespace-test... success [1m 59s]
                            grafana-test... success [3m 41s]
                        header-auth-test... success [1m 48s]
                      instance-name-test... success [2m 32s]
                             jaeger-test... success [2m 44s]
                            metrics-test... success [3m 56s]
                     null-cr-values-test... success [1m 47s]
                only-view-only-mode-test... success [2m 41s]
                             openid-test... success [1m 50s]
                     openshift-auth-test... skipped
                   os-console-links-test... skipped
                 read-certs-secrets-test... success [1m 48s]
                              roles-test... success [5m 31s]
                    rolling-restart-test... success [2m 42s]
                              token-test... success [2m 9s]
```

The times before this:
```
              accessible-namespaces-test... success [10m 3s]
     affinity-tolerations-resources-test... success [4m 16s]
                      config-values-test... success [3m 25s]
                  default-namespace-test... success [2m 30s]
                            grafana-test... success [4m 21s]
                        header-auth-test... success [2m 20s]
                      instance-name-test... success [3m 32s]
                             jaeger-test... success [3m 17s]
                            metrics-test... success [4m 26s]
                     null-cr-values-test... success [2m 17s]
                only-view-only-mode-test... success [3m 20s]
                             openid-test... success [2m 20s]
                     openshift-auth-test... skipped
                   os-console-links-test... skipped
                 read-certs-secrets-test... success [2m 20s]
                              roles-test... success [6m 43s]
                    rolling-restart-test... success [3m 20s]
                              token-test... success [2m 40s]
```

**Comment by jmazzitelli** [comment]
After all the other performance enhancements that went in, I went back and looked at the perf for turbo mode.

Running run-operator-playbook target, I get 20.6s without turbo mode, and 17.0s with turbo mode. So there is still some gain to be had, but I'm not sure if we should do it, since it is still considered experimental.

**Comment by jmazzitelli** [comment]
I'm going to close this. With the performance gains added to the operator with the PRs recently merged for this issue, and with turbo mode still listed as experimental, I think we are good for now.
